### ISSUE TITLE
Project editor validation problems

### ISSUE THREAD

**Issue by zzgvh** [akvo/akvo-rsr#3231]
Consider the following addition to IndicatorPeriod.clean():

```python
    def clean(self):
        validation_errors = {}

        if self.pk:
            orig_period = IndicatorPeriod.objects.get(pk=self.pk)

            # Don't allow an actual value to be changed when the indicator period is calculated
            if self.is_calculated() and self.actual_value != orig_period.actual_value:
                validation_errors['actual_value'] = u'%s' % \
                    _(u'It is not possible to update the actual value of this indicator period, '
                      u'because it is a calculated value. Please update the actual value through '
                      u'a new update.')

            # Don't allow some values to be changed when it is a child period
            if self.is_child_period():
                if self.indicator != orig_period.indicator:
                    validation_errors['indicator'] = u'%s' % \
                        _(u'It is not possible to update the indicator of this indicator period, '
                          u'because it is linked to a parent result.')
                if self.period_start != orig_period.period_start:
                    validation_errors['period_start'] = u'%s' % \
                        _(u'It is not possible to update the start period of this indicator, '
                          u'because it is linked to a parent result.')
                if self.period_end != orig_period.period_end:
                    validation_errors['period_end'] = u'%s' % \
                        _(u'It is not possible to update the end period of this indicator, '
                          u'because it is linked to a parent result.')

        # Don't allow a start date before an end date
        if self.period_start and self.period_end and (self.period_start > self.period_end):
            validation_errors['period_start'] = u'%s' % _(u'Period start cannot be at a later time '
                                                          u'than period end.')
            validation_errors['period_end'] = u'%s' % _(u'Period start cannot be at a later time '
                                                        u'than period end.')

        # *** New code ***
        if not self.period_start:
            validation_errors['period_start'] = u'%s' % _(u'Period start is required.')
        if not self.period_end:
            validation_errors['period_end'] = u'%s' % _(u'Period end is required.')
        # *** New code ends ***

        # TODO: add validation that prevents creating a period for a child indicator
        if validation_errors:
            raise ValidationError(validation_errors)


```

Toward the end of the method there's a section marked New code. Problem is it doesn't work. The reason for this is that the rest.views.project_editor.project_editor() view function, which is responsible for saving changes and additions made in the project editor, _saves an object one field at a time_. The object is first created but only e.g. period_start field is included as data, then when the second field of the period is found the newly created object is retrieved, the next field, e.g. period_end is added to the object and it is saved. And so on.

AFAICT the result of this incredibly convoluted way of saving data is that Model's clean() method can never check for required fields. 😲 rest.views.project_editor.project_editor() badly needs a rewrite...

**Comment by zzgvh** [comment]
I'm leaving this open as the problem remains and is blocking better data validation in the project editor.

**Comment by punchagan** [comment]
Merged into develop #3478
